fe/fe-core/src/main/java/com/starrocks/StarRocksFE.java

```diff
@@ -195,6 +205,99 @@ public static void start(String starRocksDir, String pidDir, String[] args) {
         System.exit(0);
     }
 
+    private static void handleGracefulExit() {
+        // Since the normal exit is using SIGTERM(15),
+        // so we have to choose another signal for the graceful exit, use SIGUSR1(10) here.
+        Signal.handle(new Signal("USR1"), sig -> {
+            if (canGracefulExit()) {
+                long startTime = System.nanoTime();
+                LOG.info("start to handle graceful exit");
+                GracefulExitFlag.markGracefulExit();
+
+                // transfer leader if current node is leader
+                try {
+                    transferLeader();
+                } catch (Exception e) {
+                    LOG.warn("handle graceful exit failed", e);
+                    System.exit(-1);
+                }
+
+                // Wait for queries to complete
+                try {
+                    waitForDraining();
+                } catch (Exception e) {
+                    LOG.warn("handle graceful exit failed", e);
+                    System.exit(-1);
+                }
+
+                long usedTimeMs = (System.nanoTime() - startTime) / 1000000L;
+                if (usedTimeMs < Config.min_graceful_exit_time_second * 1000L) {
+                    try {
+                        Thread.sleep(Config.min_graceful_exit_time_second * 1000L - usedTimeMs);
+                    } catch (InterruptedException ignored) {
+                    }
+                }
+
+                LOG.info("handle graceful exit successfully");
+                System.exit(0);
+            } else {
+                LOG.info("The current number of FEs that are alive cannot exit gracefully, and can only exit by force.");
+                System.exit(-1);
+            }
+        });
+    }
+
+    private static void transferLeader() throws InterruptedException {
+        if (GlobalStateMgr.getCurrentState().isLeader()) {
+            JournalWriter journalWriter = GlobalStateMgr.getCurrentState().getJournalWriter();
+            // stop journal writer
+            journalWriter.stopAndWait();
+            Journal journal = GlobalStateMgr.getCurrentState().getJournal();
+
+            // transfer leader
+            if (journal instanceof BDBJEJournal) {
+                BDBEnvironment bdbEnvironment = ((BDBJEJournal) journal).getBdbEnvironment();
+                if (bdbEnvironment != null) {
+                    // close bdb env, leader election will be triggered
+                    bdbEnvironment.close();
+                    // wait for new leader
+                    while (true) {
+                        try {
+                            InetSocketAddress address = GlobalStateMgr.getCurrentState().getHaProtocol().getLeader();
+                            LOG.info("leader is transferred to {}", address);
+                            break;
+                        } catch (Exception e) {
+                            Thread.sleep(100L);
+                        }
+                    }
```

Comment:
do we need some timeout mechanism here? what if we stuck here forever?

Reply:
It's controlled by stop_fe.sh